crates/oxc_query/src/vertex.rs

```diff
@@ -0,0 +1,319 @@
+use std::rc::Rc;
+
+use enum_as_inner::EnumAsInner;
+use oxc_ast::{ast::*, AstKind};
+use oxc_semantic::{AstNode, AstNodeId};
+use oxc_span::{GetSpan, Span};
+use url::Url;
+
+use crate::util::{expr_to_maybe_const_string, jsx_attribute_to_constant_string};
+
+#[derive(Debug, Clone, EnumAsInner)]
+pub enum Vertex<'a> {
+    ASTNode(AstNode<'a>),
+    AssignmentType(&'a BindingPatternKind<'a>),
+    Class(Rc<ClassVertex<'a>>),
+    ClassMethod(Rc<ClassMethodVertex<'a>>),
+    ClassProperty(Rc<ClassPropertyVertex<'a>>),
+    DefaultImport(&'a ImportDefaultSpecifier),
+    Expression(&'a Expression<'a>),
+    File,
+    Import(Rc<ImportVertex<'a>>),
+    Interface(Rc<InterfaceVertex<'a>>),
+    InterfaceExtend(Rc<InterfaceExtendVertex<'a>>),
+    JSXAttribute(&'a JSXAttribute<'a>),
+    JSXElement(Rc<JSXElementVertex<'a>>),
+    JSXExpressionContainer(&'a JSXExpressionContainer<'a>),
+    JSXFragment(&'a JSXFragment<'a>),
+    JSXOpeningElement(&'a JSXOpeningElement<'a>),
+    JSXSpreadAttribute(&'a JSXSpreadAttribute<'a>),
+    JSXSpreadChild(&'a JSXSpreadChild<'a>),
+    JSXText(&'a JSXText),
+    ObjectLiteral(&'a ObjectExpression<'a>),
+    PathPart(usize),
+    SearchParameter(Rc<SearchParameterVertex>),
+    Span(Span),
+    SpecificImport(&'a ImportSpecifier),
+    TypeAnnotation(Rc<TypeAnnotationVertex<'a>>),
+    Type(&'a TSType<'a>),
+    URL(Rc<Url>),
+    VariableDeclaration(Rc<VariableDeclarationVertex<'a>>),
+    ReturnStatementAST(Rc<ReturnStatementVertex<'a>>),
+}
+
+impl<'a> Vertex<'a> {
+    pub fn make_type_name(&self) -> &'static str {
```

Comment:
This should probably be an `impl<'a> Typename for Vertex<'a>`, no?

Reply:
fixed in [88ba9f7](https://github.com/web-infra-dev/oxc/pull/599/commits/88ba9f742679dcf2b55dedbf9d41cffb5f200cfb)